HW2/P5/physics.pyx

```diff
@@ -113,18 +218,61 @@ cpdef update(FLOAT[:, ::1] XY,
         #
         # SUBPROBLEM 1: parallelize this loop over 4 threads, with static
         # scheduling.
-        for i in range(count):
-            sub_update(XY, V, R, i, count, Grid, grid_spacing)
+        for i in prange(count, num_threads=n_threads,
+                        schedule='static', chunksize=chunks):
+            #sub_update(XY, V, R, i, count, Grid, grid_spacing)
+            #pass in set of locks
+            sub_update(XY, V, R, i, count, Grid, grid_spacing, locks)
 
         # update positions
         #
         # SUBPROBLEM 1: parallelize this loop over 4 threads (with static
         #    scheduling).
         # SUBPROBLEM 2: update the grid values.
-        for i in range(count):
+        for i in prange(count, num_threads=n_threads,
+                        schedule='static', chunksize=chunks):
+
+            # update grid values - delete old position for ball i from grid
+
+            if XY[i, 0]<R:
+                tempx = <int>(0)
+            elif XY[i, 0] > 1.0 - R:
+                tempx = <int>(1.0 / grid_spacing)
+            else:
+                tempx = <int>(XY[i, 0] / grid_spacing)
+
+            if XY[i, 1]<R:
+                tempy = <int>(0)
+            elif XY[i, 1] > 1.0 - R:
+                tempy = <int>(1.0 / grid_spacing)
+            else:
+                tempy = <int>(XY[i, 1] / grid_spacing)  
+
+            Grid[tempx, tempy] = -1
+
+            # new position
             for dim in range(2):
                 XY[i, dim] += V[i, dim] * t
 
+            # update index in grid values with new position for ball i;
+            # for balls projected to be off the Grid; store ball in closest
+            # Grid element
+
+            if XY[i, 0]<R:
+                tempx = <int>(0)
+            elif XY[i, 0] > 1.0 - R:
+                tempx = <int>(1.0 / grid_spacing)
+            else:
+                tempx = <int>(XY[i, 0] / grid_spacing)
+
+            if XY[i, 1]<R:
+                tempy = <int>(0)
+            elif XY[i, 1] > 1.0 - R:
+                tempy = <int>(1.0 / grid_spacing)
+            else:
+                tempy = <int>(XY[i, 1] / grid_spacing)    
+
+            Grid[tempx, tempy] = i
 
 def preallocate_locks(num_locks):
     cdef omp_lock_t *locks = get_N_locks(num_locks)
```

Comment:
I did not get a performance improvement using the locks on my computer, I am also using Hilbert and not Morton sorting, but what you get looks very consistent. Good job!